controllers/apps/components/base_rsm.go

```diff
@@ -0,0 +1,742 @@
+		updatePVCByRecreateFromStep(pvRestorePolicyStep)
+		return nil
+	}
+	if pvcQuantity := pvc.Spec.Resources.Requests[corev1.ResourceStorage]; !viper.GetBool(constant.CfgRecoverVolumeExpansionFailure) &&
+		pvcQuantity.Cmp(targetQuantity) == 1 && // check if it's compressing volume
+		targetQuantity.Cmp(*pvc.Status.Capacity.Storage()) >= 0 { // check if target size is greater than or equal to actual size
+		// this branch means we can update pvc size by recreate it
+		updatePVCByRecreateFromStep(pvPolicyRetainStep)
+		return nil
+	}
+	if pvcQuantity := pvc.Spec.Resources.Requests[corev1.ResourceStorage]; pvcQuantity.Cmp(vctProto.Spec.Resources.Requests[corev1.ResourceStorage]) != 0 {
+		// use pvc's update without anything extra
+		c.UpdateResource(newPVC, c.WorkloadVertex)
+		return nil
+	}
+	// all the else means no need to update
+
+	return nil
+}
+
+// statusUpdateTemplateRenderedConfig transform the action of rendered config by config-template from noop to update in cluster Status reconciliation
+func (c *rsmComponentBase) statusUpdateTemplateRenderedConfig() error {
+	for _, v := range ictrltypes.FindAll[*corev1.ConfigMap](c.Dag) {
+		node, _ := v.(*ictrltypes.LifecycleVertex)
+		cm, _ := node.Obj.(*corev1.ConfigMap)
+		_, ok := cm.GetLabels()[constant.CMConfigurationTypeLabelKey]
+		// generated by config-template and action is default action, so we need to change the default action
+		if ok && *node.Action == ictrltypes.NOOP {
+			newFlag := cm.Annotations[constant.CMConfigurationNewAnnotationKey]
+			if newFlag == "true" {
+				node.Action = ictrltypes.ActionCreatePtr()
+			} else {
+				node.Action = ictrltypes.ActionUpdatePtr()
+			}
+		}
+	}
+	return nil
+}
+
+func (c *rsmComponentBase) statusExpandVolume(reqCtx intctrlutil.RequestCtx, cli client.Client, txn *statusReconciliationTxn) error {
+	for _, vct := range c.runningWorkload.Spec.VolumeClaimTemplates {
+		running, failed, err := c.hasVolumeExpansionRunning(reqCtx, cli, vct.Name)
+		if err != nil {
+			return err
+		}
+		if failed {
+			txn.propose(appsv1alpha1.AbnormalClusterCompPhase, func() {
+				c.SetStatusPhase(appsv1alpha1.AbnormalClusterCompPhase, nil, "Volume Expansion failed")
+			})
+			return nil
+		}
+		if running {
+			txn.propose(appsv1alpha1.SpecReconcilingClusterCompPhase, func() {
+				c.SetStatusPhase(appsv1alpha1.SpecReconcilingClusterCompPhase, nil, "Volume Expansion failed")
+			})
+			return nil
+		}
+	}
+	return nil
+}
+
+func (c *rsmComponentBase) hasVolumeExpansionRunning(reqCtx intctrlutil.RequestCtx, cli client.Client, vctName string) (bool, bool, error) {
+	var (
+		running bool
+		failed  bool
+	)
+	volumes, err := c.getRunningVolumes(reqCtx, cli, vctName, c.runningWorkload)
+	if err != nil {
+		return false, false, err
+	}
+	for _, v := range volumes {
+		if v.Status.Capacity == nil || v.Status.Capacity.Storage().Cmp(v.Spec.Resources.Requests[corev1.ResourceStorage]) >= 0 {
+			continue
+		}
+		running = true
```

Comment:
running = true should be inited out of the loop and set properly before 'continue', failed should be set inside & outside of the loop